### PR TITLE
[SourceKit] Use an exponentially growing appending binary stream to serialize the syntax tree

### PR DESCRIPTION
**Pull request by ahoppen**
`llvm::AppendingBinaryByteStream` grows with the size of data that gets appended to it and is thus fairly inefficient if a lot of small data gets appended because it doesn't need to resize its buffer on each write. This PR introduces an output stream that doubles its internal buffer size every time it gets to small and is thus more efficient.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
> This PR looks OK to me, but how about using SmallVector<uint8_t, 0> as a storage?
> It grows exponentially, use realloc, well tested.

That's a good idea. I'll give it a shot.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: lib/Basic/ExponentialGrowthAppendingBinaryByteStream.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  if (Offset > getLength())
+    return make_error<BinaryStreamError>(stream_error_code::invalid_offset);
+
+  // Resize the internal buffer if needed.
+  uint32_t RequiredSize = Offset + Buffer.size();
+  if (RequiredSize > Data.size()) {
+    Data.resize(RequiredSize);
+  }
+
+  ::memcpy(Data.data() + Offset, Buffer.data(), Buffer.size());
```
Comment (rintaro) [inline lib/Basic/ExponentialGrowthAppendingBinaryByteStream.cpp]: Oops. It's unfortunate that `SmallVector` doesn't have equivalent method.

**Review by rintaro — APPROVED** [review]

File: lib/Basic/ExponentialGrowthAppendingBinaryByteStream.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+using namespace llvm;
+using namespace swift;
+
+Error ExponentialGrowthAppendingBinaryByteStream::readBytes(
+    uint32_t Offset, uint32_t Size, ArrayRef<uint8_t> &Buffer) {
+  if (Offset > getLength())
+    return make_error<BinaryStreamError>(stream_error_code::invalid_offset);
+
+  if (Offset + Size > getLength())
+    return make_error<BinaryStreamError>(stream_error_code::stream_too_short);
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply lib/Basic/ExponentialGrowthAppendingBinaryByteStream.cpp]: Thanks. I forgot about those methods. These are real improvements!

File: lib/Basic/ExponentialGrowthAppendingBinaryByteStream.cpp (hunk 1 above)
Comment (ahoppen) [reply lib/Basic/ExponentialGrowthAppendingBinaryByteStream.cpp]: Yeah, but still, using `SmallVector` as the underlying storage simplified things a lot.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Pull request merged**